docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,363 @@
+
+### Protocol
+
+To implement a promise module from scratch, you will have to implement the promise module protocol.
+2 variants of the protocol is provided.
+1 JSON based and 1 line based.
+
+**Note that if you are using the library, as shown above, you don't have to care / learn about the following internals of the protocol.**
+
+### Protocol header
+
+When a promise module starts, the agent sends a protocol header (single line), followed by 2 newline characters.
+The promise module responds with a protocol header (single line), followed by 2 newline characters.
+
+The protocol header contains:
+
+* name of program
+* version of program
+* version of protocol
+* space separated list of features available (optional)
+
+The header has the same syntax regardless of protocol (It is used to determine protocol).
+
+#### Example headers:
+
+Request sent by cf-agent to promise module:
+
+```
+CFEngine 3.16.0 v1
+```
+
+Response, requesting line based protocol:
+
+```
+git_promises 0.0.1 v1 line_based
+```
+
+Response, requesting JSON based protocol:
+
+```
+git_promises 0.0.1 v1 json_based
+```
+
+After the initial exchange of protocol headers, the module responds to requests from the agent, using the agreed upon protocol.
+Requests and responses follow the same format and rules, and are commonly called messages.
+For each request, there should be exactly one response.
+
+### Protocol request sequence
+
+A promise module will receive 3 types of requests (operations):
+
+1. `validate_promise` - Validate the names and types of promise attributes
+2. `evaluate_promise` - Perform changes to the system
+3. `terminate` - Shut down the module
+
+The module should not make any assumptions about the order of validation or evaluation.
+Each promise might be validated and evaluated, 0, 1, or more times, in any order.
+In general, the promise module should not maintain any state between the individual requests.
+
+The agent _might_ start multiple processes of the same module in parallel, for example for a module implementing multiple promise types.
+The module should not rely on this.
+Additionally, the same module may be started and stopped multiple times during an agent run, so don't make any assumptions that it will only run once per agent run.
+
+The promise module's responses to cf-agent requests must include `operation` and `result`.
+Promiser string and attributes are typically added for convenience.
+
+#### JSON protocol
+
+The JSON based protocol supports all aspects of CFEngine promises, including `slist`, `data` containers, and strings with newlines, tabs, etc.
+It is designed to be easy to implement in a language with JSON support.
+
+Protocol messages are separated by empty lines (double newline).
+The headers (request and response) are not JSON, but a sequence of space-separated values.
+All messages sent by cf-agent and the promise module are single line JSON-data, except:
```

Comment:
So `\n` should be used in the JSON data (strings) for newlines?

Reply:
Added note about this.